### ISSUE TITLE
Callback to iterator utility: cb_to_iter

### ISSUE THREAD

**Issue by frankier** [more-itertools/more-itertools#441]
Thought I would ask about this one here since it's something I keep using, and is completely to do with iterators but I'm not sure if it might be a bit controversial/out of scope.

Some libraries (e.g. HDF5py, lxml) do iteration via callbacks. It seems to be particularly common in fairly straight forward wrappers of C/C++ APIs. The idea of converting it to an iterator has come up before https://stackoverflow.com/questions/9968592/turn-functions-with-a-callback-into-python-generators . I myself have used this repeatedly (I have been copy pasting it between projects) so that I can get something out I can start applying iterator combinators on.

Here is the straw man. 

```
class AbortThread(BaseException):
    pass


def cb_to_iter(f):
    def iter(*args, **kwargs):
        from threading import Thread
        from queue import Queue

        q = Queue()
        job_done = object()
        abort = False

        def cb(*cb_args, **cb_kwargs):
            if cb_kwargs:
                cb_args = cb_args + (cb_kwargs,)
            if len(cb_args) == 1:
                q.put(cb_args[0])
            else:
                q.put(cb_args)
            q.join()
            if abort:
                raise AbortThread

        def task(*args, **kwargs):
            try:
                f(*(args + (cb,)), **kwargs)
                q.put(job_done)
            except AbortThread:
                pass

        thread = Thread(target=task, args=args, kwargs=kwargs)
        thread.start()

        try:
            while True:
                next_item = q.get(True)
                if next_item is job_done:
                    break
                yield next_item
                q.task_done()
        finally:
            abort = True
            try:
                q.task_done()
            except ValueError:
                pass
            thread.join(5)

    return iter
```

One big question mark here is: is it okay to start spawning threads? The main justification has to be that it's the only way to achieve this, and this is the only way to start applying iterator combinators when faced with this type of API short of just collecting everything into a list.

It would also be possible to use an async iterator, but that's a whole different kettle of fish.

**Comment by bbayles** [comment]
Definitely a useful thing, though the way one has to do this is quite... un-aesthetic. Can you paste some examples of your code in action, e.g. with the scipy `fmin` from SO?

The `thread.join(5)` in the `finally` is unfortunate, but I suppose it's either that or risk leaking resources?

**Comment by bbayles** [comment]
Here's my take:

<details><summary>callback_iter</summary>

```python
def callback_iter(func, yield_result=False, wait_seconds=0.1):
    with ThreadPoolExecutor(max_workers=1) as executor:
        q = Queue()
        future = executor.submit(func, callback=lambda x: q.put(x))
        while True:
            try:
                item = q.get(timeout=wait_seconds)
            except Empty:
                pass
            else:
                q.task_done()
                yield item

            if future.done():
                result = future.result()
                break

        while True:
            try:
                item = q.get_nowait()
            except Empty:
                break
            else:
                q.task_done()
                yield item

        q.join()

        if yield_result:
            yield result
```

</details>

In action:

```python
def output_maker(cb=None):
    for i in ['1', '2', '3', '4']:
        if func:
            cb(i)
    return '5'

func = lambda callback=None: output_maker(cb=callback)
it = callback_iter(func, yield_result=True)
next(it)  # '1'
next(it)  # '2'
list(it)  # ['3', '4', '5']

```

One issue with this (and maybe with yours?) occurs when the wrapped function sends multiple arguments to the callback. For example:

```python
from more_itertools.more import callback_iter
from timeit import Timer


func = Timer().autorange
it = callback_iter(func)
next(it)  # TypeError: <lambda>() takes 1 positional argument but 2 were given
```

I could fix this by doing this:

```python
        future = executor.submit(func, callback=lambda *x: q.put(x))
```

but then the yielded items are `tuple`s.

```python
from more_itertools.more import callback_iter
from timeit import Timer


func = Timer().autorange
it = callback_iter(func)
next(it)  (1, 7.109993021003902e-07)
```

What do you think?

**Comment by frankier** [comment]
Your version is simpler, but I'm not sure if it's worth removing the use of the sentinel object to indicate the producer thread is finished and replacing it with busy waiting with `wait_seconds`. The reason is that if `wait_seconds` is too short, we will spend a substantial amount of time just checking future.done(), but if it's too long we could wait too long when the queue if already finished.

Another thing is if a user doesn't consume the whole iterator, it looks like your version will run the function in the thread to completion rather than trying to abort it using an exception. I'm not sure what's best here, but it might be better your way and just tell people that they should consume the whole iterator. On the other hand, even if the user intends to consume the whole iterator, if there's an exception in user code they will see the exception printed, but Python will not exit since it will be waiting for the producer thread to complete.

I do think using `with ThreadPoolExecutor` would be good, since it gives us extra assurance that we're not leaking threads.

The interface of taking a callback keyword argument and expecting the user the dispatch using a lambda seems nicer.

I think always yielding tuples is fine. At least it's consistent.

At the risk of complicating this further again, another consideration is whether there should be the option of chunking. This is because queueing/dequeuing from the queue can be a bottleneck since it requires a synchronised context switch for each item. If there are many small items this is not good. Of course it would be nice if we could use combinators for this, but the chunking should take place in the producer thread. An option to use fixed chunk sizes with a default of `1` would probably do the job.

**Comment by bbayles** [comment]
> The reason is that if wait_seconds is too short, we will spend a substantial amount of time just checking future.done()

Hmm, `future.done()` doesn't block. So in this loop the purpose of `wait_seconds` is to avoid busy-waiting on the queue to have an item.

<details><summary>Queue read loop</summary>

```python
        while True:
            try:
                item = q.get(timeout=wait_seconds)
            except Empty:
                pass
            else:
                q.task_done()
                yield item

            if future.done():
                result = future.result()
                break
```
</details>


> Another thing is if a user doesn't consume the whole iterator, it looks like your version will run the function in the thread to completion rather than trying to abort it using an exception. On the other hand, even if the user intends to consume the whole iterator, if there's an exception in user code they will see the exception printed, but Python will not exit since it will be waiting for the producer thread to complete.

That was a simplification, yeah. I think this [version](https://github.com/more-itertools/more-itertools/blob/711659d373a9fae50dbd06c263603fd510215a09/more_itertools/more.py#L3252) behaves reasonably well if you don't consume everything. If you never get past [this point](https://github.com/more-itertools/more-itertools/blob/711659d373a9fae50dbd06c263603fd510215a09/more_itertools/more.py#L3312), the thread will run to completion, sticking all output in the queue.

Re: chunking, I think that's beyond what I'd put in this library.
